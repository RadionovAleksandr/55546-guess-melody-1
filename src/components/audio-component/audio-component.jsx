import React from 'react';
import propTypes from 'prop-types';

export const snapshotURL = `void://for-snapshot`;

export default class AudioComponent extends React.PureComponent {
  render() {
    const {isLoading, isPlaying, renderAudio, onPlayButtonClick} = this.props;

    return (
      <>
        <button
          className={`track__button track__button--${isPlaying ? `pause` : `play`}`}
          type="button"
          disabled={isLoading}
          onClick={onPlayButtonClick}
        />
        <div className="track__status">
          {renderAudio()}
        </div>
      </>
    );
  }
}

AudioComponent.propTypes = {
  isLoading: propTypes.bool.isRequired,
  isPlaying: propTypes.bool.isRequired,
  onPlayButtonClick: propTypes.func.isRequired,
  renderAudio: propTypes.func.isRequired,
  src: propTypes.string.isRequired,
};


// class AudioComponent extends React.PureComponent {
//   constructor(props) {
//     super(props);
//     if (props.src === snapshotURL) {
//       this._audioRef = {
//         current: {
//           play() {},
//           pause() {}
//         }
//       };
//     } else {
//       this._audioRef = React.createRef();
//     }

//     this.state = {
//       progress: 0,
//       isLoading: true,
//       isPlaying: props.isPlaying,
//     };

//     this._onPlayButtonClick = this._onPlayButtonClick.bind(this);
//   }

//   render() {
//     const {isLoading, isPlaying} = this.state;

//     return (
//       <React.Fragment>
//         <button
//           className={`track__button track__button--${isPlaying ? `pause` : `play`}`}
//           type="button"
//           disabled={isLoading}
//           onClick={this._onPlayButtonClick}
//         />
//         <div className="track__status">
//           <audio
//             ref={this._audioRef}
//           />
//         </div>
//       </React.Fragment>
//     );
//   }

//   componentDidMount() {
//     const {src} = this.props;

//     if (src === snapshotURL) {
//       return;
//     }

//     const audio = this._audioRef.current;

//     audio.src = src;

//     audio.oncanplaythrough = () => this.setState({
//       isLoading: false,
//     });

//     audio.onplay = () => {
//       this.setState({
//         isPlaying: true,
//       });
//     };

//     audio.onpause = () => this.setState({
//       isPlaying: false,
//     });

//     audio.ontimeupdate = () => this.setState({
//       progress: audio.currentTime
//     });
//   }

//   componentDidUpdate() {
//     const audio = this._audioRef.current;

//     audio[this.props.isPlaying ? `play` : `pause`]();
//   }

//   componentWillUnmount() {
//     const audio = this._audioRef.current;

//     audio.oncanplaythrough = null;
//     audio.onplay = null;
//     audio.onpause = null;
//     audio.ontimeupdate = null;
//     audio.src = ``;
//   }

//   _onPlayButtonClick() {
//     this.props.onPlayButtonClick();
//     this.setState({isPlaying: !this.state.isPlaying});
//   }
// }

// AudioComponent.propTypes = {
//   isPlaying: propTypes.bool.isRequired,
//   onPlayButtonClick: propTypes.func.isRequired,
//   src: propTypes.string.isRequired,
// };

// export default AudioComponent;
