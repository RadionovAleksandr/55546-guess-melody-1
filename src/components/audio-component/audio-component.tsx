import * as React from 'react';

export const snapshotURL = `void://for-snapshot`;

interface Props {
  isLoading: boolean,
  isPlaying: boolean,
  onPlayButtonClick: () => void,
  renderAudio: () => React.ReactElement,
}

export default class AudioComponent extends React.PureComponent<Props, null> {
  render() {
    const {isLoading, isPlaying, renderAudio, onPlayButtonClick} = this.props;

    return (
      <>
        <button
          className={`track__button track__button--${isPlaying ? `pause` : `play`}`}
          type="button"
          disabled={isLoading}
          onClick={onPlayButtonClick} />
        <div className="track__status">{renderAudio()}</div>
      </>
    );
  }
}
