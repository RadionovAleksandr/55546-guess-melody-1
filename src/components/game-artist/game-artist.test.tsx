import * as React from 'react';
import * as renderer from 'react-test-renderer';

import GameArtist from './game-artist';
import {snapshotURL} from '../audio-component/audio-component';

import {Type} from "../../types";

const mock = {
  questions: [
    {
      type: Type.GENRE,
      genre: `rock`,
      answers: [
        {
          src: snapshotURL,
          genre: `rock`,
        },
        {
          src: snapshotURL,
          genre: `blues`,
        },
        {
          src: snapshotURL,
          genre: `jazz`,
        },
        {
          src: snapshotURL,
          genre: `rock`,
        },
      ],
    },
    {
      type: Type.ARTIST,
      song: {
        artist: `Jim Beam`,
        src: snapshotURL,
      },
      answers: [
        {
          picture: `path.jpg`,
          artist: `John Snow`,
        },
        {
          picture: `path.jpg`,
          artist: `Jack Daniels`,
        },
        {
          picture: `path.jpg`,
          artist: `Jim Beam`,
        },
      ],
    }
  ],
};

describe(`Test GameArtist`, () => {
  it(`GameArtist is rendered correctly`, () => {
    const tree = renderer.create(<GameArtist
      renderPlayer={jest.fn()}
      onAnswer={jest.fn()}
      question={mock.questions[0]}
    />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
