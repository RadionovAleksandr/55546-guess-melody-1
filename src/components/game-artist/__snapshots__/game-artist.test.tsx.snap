// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test GameArtist GameArtist is rendered correctly 1`] = `
<section
  className="game__screen"
>
  <h2
    className="game__title"
  >
    Кто исполняет эту песню?
  </h2>
  <div
    className="game__track"
  />
  <form
    className="game__artist"
  >
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="artist-0"
        name="answer"
        onClick={[Function]}
        type="radio"
        value="artist-0"
      />
      <label
        className="artist__name"
        htmlFor="artist-0"
      >
        <img
          className="artist__picture"
        />
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="artist-1"
        name="answer"
        onClick={[Function]}
        type="radio"
        value="artist-1"
      />
      <label
        className="artist__name"
        htmlFor="artist-1"
      >
        <img
          className="artist__picture"
        />
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="artist-2"
        name="answer"
        onClick={[Function]}
        type="radio"
        value="artist-2"
      />
      <label
        className="artist__name"
        htmlFor="artist-2"
      >
        <img
          className="artist__picture"
        />
      </label>
    </div>
    <div
      className="artist"
    >
      <input
        className="artist__input visually-hidden"
        id="artist-3"
        name="answer"
        onClick={[Function]}
        type="radio"
        value="artist-3"
      />
      <label
        className="artist__name"
        htmlFor="artist-3"
      >
        <img
          className="artist__picture"
        />
      </label>
    </div>
  </form>
</section>
`;
